README
NAME
    MooX::Lsub - Very shorthand syntax for bulk lazy builders

VERSION
    version 0.002001

SYNOPSIS
      use MooX::Lsub;

      # Shorthand for
      # has foo => ( is => ro =>, lazy => 1, builder => '_build_foo' );
      # sub _build_foo { "Hello" }

      lsub foo => sub { "Hello" };

DESCRIPTION
    I often want to use a lot of lazy build subs to implement some plumbing,
    with scope to allow it to be overridden by people who know what they're
    doing with an injection library like Bread::Board.

    Usually, the syntax of "Class::Tiny" is what I use for such things.

      use Class::Tiny {
        'a' => sub { },
        'b' => sub { },
      };

    Etc.

    But switching things to Moo means I usually have to get much uglier, and
    repeat myself a *lot*.

    So this module exists as a compromise.

    Additionally, I always forgot to declare "use Moo 1.000008" which was
    the first version of "Moo" where "builder => sub" worked, and I would
    invariably get silly test failures in smokers as a consequence.

    This module avoids such problem entirely, and is tested to work with
    "Moo 0.009001".

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for MooX-LSub

0.002002 2017-03-03T14:18:17Z fac06c0

0.002001 2017-03-03T12:23:13Z 6a03f10
 [Bugfix]
 - Avoid test failures on 5.26 w/ -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/MooX-Lsub/issues/2

 [Dependencies::Stats]
 - Dependencies changed since 0.002000, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - test: +2 ↓1 -1 (recommends: +1 ↑2 ↓1)

 [Misc]
 - use Test::Needs instead of Test::Require for faster skipping when deps aren't installed ( avoids spurious load of
   Test::More and reams of Test2 for a simple noop )

0.002000 2014-08-30T11:24:12Z
 [00 Minor]
 - Switch to Sub::Util

 [Dependencies::Stats]
 - Dependencies changed since 0.001003, see misc/*.deps* for details
 - develop: +2 ↑1 -1 (suggests: ↑2)
 - runtime: (recommends: +1 -1)

 [Meta]
 - use kentnl for author

0.001003 2014-08-15T23:15:29Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests augmented.
 - whitespace reduced.

 [Dependencies::Stats]
 - Dependencies changed since 0.001002, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: -4 (recommends: ↑1 -2)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update Test::ReportPrereqs

0.001002 2014-07-21T05:05:52Z
 [Bugfix]
 - Name the builder sub where possible

 [Dependencies::Stats]
 - Dependencies changed since 0.001001, see misc/*.deps* for details
 - develop: +1 (suggests: ↑2)
 - runtime: +1 (recommends: +1)
 - test: +2 ↑1 (recommends: +1)

0.001001 2014-07-08T07:29:43Z
 [Bugfix]
 - Test failures erroneously if Moo is not installed.

0.001000 2014-07-07T06:33:32Z
 - First version.


lib/MooX/Lsub.pm
use 5.008;    # utf8
use strict;
use warnings;
use utf8;

package MooX::Lsub;

our $VERSION = '0.002002';

# ABSTRACT: Very shorthand syntax for bulk lazy builders

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Eval::Closure qw(eval_closure);
use Carp qw(croak);
## no critic (Capitalization,ProhibitConstantPragma,RequireCheckingReturnValueOfEval);
use constant can_haz_subname => eval { require Sub::Util };

## no critic (TestingAndDebugging::ProhibitNoStrict)
sub _get_sub {
  my ( undef, $target, $subname ) = @_;
  no strict 'refs';
  return \&{ $target . q[::] . $subname };
}

sub _set_sub {
  my ( undef, $target, $subname, $code ) = @_;
  no strict 'refs';
  *{ $target . q[::] . $subname } = $code;
  return;
}

sub _set_sub_named {
  my ( undef, $target, $subname, $code ) = @_;
  no strict 'refs';
  *{ $target . q[::] . $subname } = can_haz_subname ? Sub::Util::set_subname( $target . q[::] . $subname, $code ) : $code;
  return;
}
## use critic
#
sub import {
  my ( $class, @args ) = @_;
  my $target = caller;
  my $has = $class->_get_sub( $target, 'has' );

  croak "No 'has' method in $target. Did you forget to import Moo(se)?" if not $has;

  my $lsub_code = $class->_make_lsub(
    {
      target  => $target,
      has     => $has,
      options => \@args,
    },
  );

  $class->_set_sub( $target, 'lsub', $lsub_code );

  return;
}

sub _make_lsub_code {
  my ( $class, $options ) = @_;
  ## no critic (ValuesAndExpressions::RequireInterpolationOfMetachars)
  my $nl   = qq[\n];
  my $code = 'sub($$) {' . $nl;
  $code .= q[ package ] . $class . q[; ] . $nl;
  $code .= q[ my ( $subname, $sub , @extras ) = @_; ] . $nl;
  $code .= q[ if ( @extras ) { ] . $nl;
  $code .= q[   croak "Too many arguments to 'lsub'. Did you misplace a ';'?"; ] . $nl;
  $code .= q[ } ] . $nl;
  $code .= q[ if ( not defined $subname or not length $subname or ref $subname ) { ] . $nl;
  $code .= q[   croak "Subname must be defined + length + not a ref"; ] . $nl;
  $code .= q[ } ] . $nl;
  $code .= q[ if ( not 'CODE' eq ref $sub ) { ] . $nl;
  $code .= q[   croak "Sub must be a CODE ref"; ] . $nl;
  $code .= q[ } ] . $nl;
  $code .= q[ $class->_set_sub_named($target, "_build_" . $subname , $sub ); ] . $nl;
  $code .= q[ package ] . $options->{'target'} . q[; ] . $nl;
  $code .= q[ return $has->( ] . $nl;
  $code .= q[   $subname, ] . $nl;
  $code .= q[   ( ] . $nl;
  $code .= q[     is => 'ro', ] . $nl;
  $code .= q[     lazy => 1, ] . $nl;
  $code .= q[     builder => '_build_' . $subname, ] . $nl;
  $code .= q[   ) ] . $nl;
  $code .= q[ ); ] . $nl;
  $code .= q[}] . $nl;
  ## use critic
  return $code;
}

sub _make_lsub {
  my ( $class, $options ) = @_;

  my $code = $class->_make_lsub_code($options);

  my $sub = eval_closure(
    source      => $code,
    environment => {
      ## no critic (ValuesAndExpressions::RequireInterpolationOfMetachars)
      '$class'  => \$class,
      '$has'    => \$options->{'has'},
      '$target' => \$options->{'target'},
      ## use critic
    },
  );
  return $sub;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

MooX::Lsub - Very shorthand syntax for bulk lazy builders

=head1 VERSION

version 0.002002

=head1 SYNOPSIS

  use MooX::Lsub;

  # Shorthand for
  # has foo => ( is => ro =>, lazy => 1, builder => '_build_foo' );
  # sub _build_foo { "Hello" }

  lsub foo => sub { "Hello" };

=head1 DESCRIPTION

I often want to use a lot of lazy build subs to implement some plumbing, with scope to allow
it to be overridden by people who know what they're doing with an injection library like Bread::Board.

Usually, the syntax of C<Class::Tiny> is what I use for such things.

  use Class::Tiny {
    'a' => sub { },
    'b' => sub { },
  };

Etc.

But switching things to Moo means I usually have to get much uglier, and repeat myself a *lot*.

So this module exists as a compromise.

Additionally, I always forgot to declare C<use Moo 1.000008> which was the first version of C<Moo> where
C<< builder => sub >> worked, and I would invariably get silly test failures in smokers as a consequence.

This module avoids such problem entirely, and is tested to work with C<Moo 0.009001>.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Very shorthand syntax for bulk lazy builders",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "MooX-Lsub",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "MooX::Lsub",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Eval::Closure" => 0,
    "constant" => 0,
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "Test::Needs" => 0
  },
  "VERSION" => "0.002002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Eval::Closure" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => 0,
  "Test::Needs" => 0,
  "constant" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
